[DR-2425] Add PatchDataset, PatchSnapshot endpoints for updating passport identifiers

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-2425

**Code Changes**

Added two new endpoints to synchronously modify passport identifiers:

- [PatchDataset](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/datasets/patchDataset) - can set or override PHS ID
- [PatchSnapshot](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/snapshots/patchSnapshot) - can set or override consent code

In the future, we will expand on these endpoints to support the modification of other resource fields (ex. dataset description).

As we're following a [JSON merge patch](https://www.rfc-editor.org/rfc/rfc7396) approach in our response body format, we can't unset fields via these endpoints.  To work around this, callers can set them to empty strings.  We've changed our "passportabililty" check accordingly.

Because we can modify objects cached in `DrsService`,  I've expanded ga4gh endpoint descriptions to warn of our 15 minute local cache expiration.  A user might be confused by the following experience:
- Call `OptionsObject`: see that their object was missing the necessary passport identifiers.
- Modify dataset and/or snapshot to populate passport identifiers.
- Call `OptionsObject`: no immediate change in response, the service has the old objects cached.

A more sophisticated approach would involve cache eviction, but this was out of scope and would likely involve rethinking our caching strategy.

**Manual Verification**

My [dev environment](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#) reflects these changes.
Using the following resources:

- [Dataset](https://jade-ok.datarepo-dev.broadinstitute.org/datasets/79ef03d8-161a-4143-bb09-c00cf7c11570)
- [Snapshot](https://jade-ok.datarepo-dev.broadinstitute.org/snapshots/6510b127-c226-46a1-ac12-c5f2383cbd08) whose sole row has a field with DRS object ID `v1_6510b127-c226-46a1-ac12-c5f2383cbd08_935baf22-8428-4cd1-b229-9e2a9fb2c2d7`

I used my dev Swagger to modify the passport identifiers, and verified that `OptionsObject` reflected the new state of the dataset and snapshot.

I should have added DataRepoAdmins@dev.test.firecloud.org to all necessary resources in case anyone would like to play around!

okotsopoulos committed:
Add PatchDataset endpoint for updating a dataset's PHS ID

okotsopoulos committed:
Spotless (sad that it makes the comment harder to read)

okotsopoulos committed:
Prefer less specific parameter names

okotsopoulos committed:
spotlessJavaApply

okotsopoulos committed:
Add PatchSnapshot endpoint for updating a snapshot's consent code

okotsopoulos committed:
Passportability check now requires PHS ID and consent code to not be blank

Previously we performed null checks on these values.
We can't null out values in our patch requests but we can set them to empty strings.

okotsopoulos committed:
Check appropriate IamAction when verifying PatchSnapshot auth

okotsopoulos committed:
Add dataset and snapshot DAO unit tests

okotsopoulos committed:
Add Swagger ga4gh warnings re: cache expiration

okotsopoulos commented on src/main/java/bio/terra/service/dataset/DatasetDao.java:
```diff
@@ -703,6 +704,25 @@ public DatasetSummary mapRow(ResultSet rs, int rowNum) throws SQLException {
+   * @return whether the dataset record was updated
+   */
+  @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+  public boolean patch(UUID id, DatasetPatchRequestModel patchRequest) {
```
One thing that I haven't added yet is any kind of logging, debug or otherwise. I welcome suggestions or examples on what might be helpful (and can also talk to Ruchi about audit trail requirements that may come down the pike).

snf2ye replied:
I think an audit log would be helpful to indicate the updates made (Just a logger.info() statement). Although it might be tricky to report which fields are actually updated since it relies on COALESCE to determine whether the field is null in the patch request.

okotsopoulos replied:
I added in logging on successful patches, I kept the logging statement simple in that I'm logging the stringified patch request.  While it doesn't exclude any null fields, I think it meets our needs for auditing for now.  Here's an example log from a successful snapshot patch:

```
{"timestampSeconds":1653318344,"timestampNanos":124000000,"severity":"INFO","message":"Snapshot 6510b127-c226-46a1-ac12-c5f2383cbd08 patched with class SnapshotPatchRequestModel {\n    consentCode: c03\n}","serviceContext":{"service":null,"version":null},"context":"default","thread":"http-nio-8080-exec-5","logger":"bio.terra.service.snapshot.SnapshotDao","logging.googleapis.com/sourceLocation":{"file":"bio/terra/service/snapshot/SnapshotDao.java","line":701,"function":"bio.terra.service.snapshot.SnapshotDao.patch"},"logging.googleapis.com/trace":"projects/broad-jade-dev/traces/8f0a46e8b13f1756181b044de2d92210","logging.googleapis.com/spanId":"38c45d15a2f2c01e","logging.googleapis.com/trace_sampled":false,"traceId":"616f81b02ad57bba","spanId":"616f81b02ad57bba","requestId":"6WYvDwyW"}
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetDao.java:
```diff
@@ -703,6 +704,25 @@ public DatasetSummary mapRow(ResultSet rs, int rowNum) throws SQLException {
+   */
+  @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+  public boolean patch(UUID id, DatasetPatchRequestModel patchRequest) {
+    String sql = "UPDATE dataset SET phs_id = COALESCE(:phs_id, phs_id) WHERE id = :id";
```
I'm curious about this use of `COALESCE`. My understanding is that `COALESCE(x, y, z)` returns the first non-null object. So if `patchRequest.getPhsId()` is `null`, then this would be a no-op. If that's the desired behavior, could you get the same result by skipping the `update()` call when `patchRequest.getPhsId()` is `null`?

okotsopoulos replied:
Thanks for asking.  That's my understanding of `COALESCE` as well.  I like using it here because we'll soon expand these endpoints to allow other elements of the dataset / snapshot to be modified, this way we'll keep the existing values for any `patchRequest` elements left unspecified.

A concrete example, if/when we allow dataset descriptions to be modified:
```
UPDATE dataset SET
  phs_id = COALESCE(:phs_id, phs_id),
  description = COALESCE(:description, description)
WHERE id = :id
```

A user could then overwrite the description without modifying the PHS ID.

pshapiro4broad replied:
That makes sense, I didn't know that there's a future need for updating other elements in this API.

samanehsan commented on src/main/java/bio/terra/app/controller/DatasetsApiController.java:
```diff
@@ -176,6 +178,15 @@ public ResponseEntity<JobModel> deleteDataset(@PathVariable("id") UUID id) {
+      UUID id, DatasetPatchRequestModel patchRequest) {
+    AuthenticatedUserRequest userReq = getAuthenticatedInfo();
+    iamService.verifyAuthorization(
+        userReq, IamResourceType.DATASET, id.toString(), IamAction.MANAGE_SCHEMA);
```
Both dataset [stewards](https://github.com/broadinstitute/sam/blob/develop/src/main/resources/reference.conf#L1060) and [custodians](https://github.com/broadinstitute/sam/blob/develop/src/main/resources/reference.conf#L1063) have the `mangage_schema` action in SAM. I think it makes sense that a data custodian would want the ability to PATCH dataset information (especially when we include more fields) but updating the PHS Id indirectly affects who has access to the data.

Right now only dataset stewards and admins are allowed to update who has access to a dataset in SAM (they have the `share_policy::<role>` actions). This change would allow custodians to indirectly change who has read access, but I think that's okay because they still can't add other custodians or stewards. What do you think?

okotsopoulos replied:
I'm glad that you picked up on this nuance.  Thanks as well for the links to the sam configurations!

What could work here is conditionally checking for some `share_policy` authorization in addition to the `manage_schema` check, in cases where we are modifying passport identifiers.  I may ping you about this to chat further though: I see we register this authorization as `IamAction.SHARE_POLICY` without the role qualifications, and the top-level `IamAction` comment reads as follows:

> // NOTE: these action enums must have exactly the same text as the Sam action name.

samanehsan replied:
That's a good point! I did a quick search and we're not using `SHARE_POLICY`, `READ_POLICY`, or `READ_POLICIES` anywhere (and `ALTER_POLICY` is only used in tests). I'm guessing we left out checking these actions on our end because SAM already knows who can do these actions. So we would need to update our actions enum with the more specific share policies.

okotsopoulos replied:
I've made this change and tested it out on my dev environment:

- I added my okotsopo@broadinstitute.org account as a custodian but not a steward of the dataset.  I can still call `PatchDataset` but when I specify a PHS ID change, I get an error message indicating that I'm missing `share_policy::steward`.
- My test account is still a steward of the dataset and can still modify its PHS ID.

I welcome any further feedback especially around the `IamActions` changes, since I only added the `share_policy::steward` action (and I’m not convinced that’s the appropriate share policy to check for).

samanehsan replied:
Thanks for making these changes! We actually want to check if the user has the `share_policy::reader` action because that means they are allowed to add readers to the dataset

snf2ye replied:
This is definitely an interesting case - Thanks for thinking about this! I don't think I would have caught it.

Two things to think about:
With both MANAGE_SCHEMA and SHARE_POLICY_STEWARD, if I'm understanding correctly, we're saying that a user can alter the PHS ID in two cases:
1. Have steward role
2. Have Custodian AND Admin Role (custodian provides manage schema permission and admin provides the share policy permission)

I think 2 is problematic. As far as I know, we don't usually allow combining roles to get to the right permissions. It would be better to add a new SAM action and give it direct access to the particular roles we've determined. 

I also think this will be an issue when we add additional fields to the update. For example, with the "additional properties" field we're about to add to the dataset patch, the user doesn't need the share_policy permission. Therefore, Custodians will be blocked from making the update, when they really should have permission.

okotsopoulos replied:
@samanehsan

RE: `share_policy::reader`:

I thought that role sounded right at first, but this is a share policy associated with snapshots and not datasets.  I confirmed this manually by checking for `share_policy::reader` on my test dataset in [sam dev API](https://sam.dsde-dev.broadinstitute.org/#/Resources/getPolicyV2), it returned false.

Here are the share policies granted to a [dataset steward](https://github.com/broadinstitute/sam/blob/develop/src/main/resources/reference.conf#L1060):
```
"share_policy::steward", "share_policy::custodian", "share_policy::ingester", "share_policy::snapshot_creator"
```

okotsopoulos replied:
@snf2ye 

RE: stacking roles to complete needed permissions:

This counterexample helps explain what felt like a funny smell :) initially I didn't realize what control we had over SAM actions.  I'd like to talk to you about what's needed to create a new one, etc.

RE: Adding additional fields to update:

This probably isn't relevant now, but I don't think this would pose a problem, as the share_policy permission is conditionally checked.  We only check it when modifying a passport identifier.  For "additional properties" patching we would only check for manage_schema permission.

samanehsan replied:
Oh! You're completely right @okotsopoulos! I mixed up dataset and snapshot policies 🤦‍♀️ That's probably why there was only a `SHARE_POLICY` defined then since the steward is currently the only role that can share datasets. In that case, we should check if a user has any of the `share_policy` actions for datasets since all of the policies grant read access. I think we can exclude `ingester`, since it is a deprecated policy (I'll add that to our SAM spring-cleaning list).

samanehsan replied:
Hmm I see Shelby's point now...updating the SAM config with a new action for patching the PHS ID and giving it to the dataset steward role would fix this.

snf2ye replied:
>  don't think this would pose a problem, as the share_policy permission is conditionally checked.

Ah, ok! Makes sense!

okotsopoulos replied:
Okay, this should be resolved!

New SAM action `update_passport_identifier` for dataset and snapshot stewards only is live on dev.  We now conditionally check for this action if updating passport identifiers on datasets or snapshots.  I verified this behavior on my dev environment.

samanehsan commented on src/test/java/bio/terra/service/dataset/DatasetDaoTest.java:
```diff
@@ -834,4 +835,39 @@ public void columnOrderCreationTest() throws Exception {
+        equalTo(phsIdOverride));
+
+    datasetDao.patch(datasetId, new DatasetPatchRequestModel());
+    assertThat(
```
Could we add another case here where the PHS ID is "unset" by changing it to an empty string (same with the snapshot consent code test)?

okotsopoulos replied:
Added!

samanehsan commented:
Thank you for deploying these changes to your test environment and creating test data! ❤️

okotsopoulos committed:
@sehsan PR feedback: test patching with empty string

okotsopoulos committed:
@sehsan PR feedback: conditionally check for share privileges when modifying a dataset's passport identifier

okotsopoulos committed:
spotless

samanehsan commented on src/main/java/bio/terra/service/dataset/DatasetService.java:
```diff
@@ -256,6 +258,22 @@ public String delete(String id, AuthenticatedUserRequest userReq) {
+   * @param patchRequest updates to merge with an existing dataset
+   * @return IAM actions needed to apply the requested patch
+   */
+  public List<IamAction> iamActions(DatasetPatchRequestModel patchRequest) {
```
I think giving this a more specific name would be helpful, like `iamPatchDatasetActions`

okotsopoulos replied:
Altered!

okotsopoulos committed:
@samanehsan @snf2ye PR feedback

- Conditionally check for new SAM action update_passport_identifier when patching resources
- More specific method naming

okotsopoulos committed:
@snf2ye PR feedback: add logging on successful patch

samanehsan reviewed (approved):
Lgtm! :octocat:

pshapiro4broad commented on src/main/java/bio/terra/app/controller/DatasetsApiController.java:
```diff
@@ -176,6 +178,16 @@ public ResponseEntity<JobModel> deleteDataset(@PathVariable("id") UUID id) {
+  public ResponseEntity<DatasetSummaryModel> patchDataset(
+      UUID id, DatasetPatchRequestModel patchRequest) {
+    AuthenticatedUserRequest userReq = getAuthenticatedInfo();
+    for (IamAction action : datasetService.patchDatasetIamActions(patchRequest)) {
+      iamService.verifyAuthorization(userReq, IamResourceType.DATASET, id.toString(), action);
+    }
```
This will make a separate SAM call for each action. It looks TDR doesn't use it yet, but there's a SAM API that returns all actions for a user for a resource (`resourceActionsV2`). (TDR calls this via `IamProviderInterface.hasActions` but the actions aren't returned.) It would be better to only make a single SAM call here, if possible.

okotsopoulos replied:
Thanks for the tip, I will take a closer look.

okotsopoulos replied:
After looking at the code, my plan is to merge this PR as it stands and handle this consolidation separately in a new ticket since the changes may be a touch involved.  The endpoint will not be called frequently so we won't clobber SAM in the interim.

https://broadworkbench.atlassian.net/browse/DR-2607

pshapiro4broad commented on src/main/java/bio/terra/service/auth/iam/IamAction.java:
```diff
@@ -33,16 +33,22 @@ public enum IamAction {
+  private final String samActionName;
+
+  IamAction() {
+    this.samActionName = StringUtils.lowerCase(name());
```
Since `name` can't be `null` there's no hazard with calling the `String` API directly
```suggestion
    this.samActionName = name().toLowerCase();
```

okotsopoulos committed:
Streamline IamAction.samActionName assignment

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>
